nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/excel/ExcelReader.java

```diff
@@ -0,0 +1,193 @@
+                return delimitedSheets;
+            } else {
+                if (logger != null) {
+                    logger.debug("Excel document was parsed but no sheets with the specified desired names were found.");
+                }
+            }
+        }
+
+        return new String[0];
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> properties = new ArrayList<>(super.getSupportedPropertyDescriptors());
+        properties.add(DESIRED_SHEETS);
+        properties.add(FIRST_ROW_NUM);
+        properties.add(DateTimeUtils.DATE_FORMAT);
+        properties.add(DateTimeUtils.TIME_FORMAT);
+        properties.add(DateTimeUtils.TIMESTAMP_FORMAT);
+
+        return properties;
+    }
+
+    @Override
+    protected SchemaAccessStrategy getSchemaAccessStrategy(final String allowableValue, final SchemaRegistry schemaRegistry, final PropertyContext context) {
+        if (allowableValue.equalsIgnoreCase(HEADER_DERIVED.getValue())) {
+            return new ExcelHeaderSchemaStrategy(context);
+        } else if (allowableValue.equalsIgnoreCase(SchemaInferenceUtil.INFER_SCHEMA.getValue())) {
+            final RecordSourceFactory<Row> sourceFactory = (variables, in) -> new ExcelRecordSource(in, context, variables);
+            final SchemaInferenceEngine<Row> inference = new ExcelSchemaInference(new TimeValueInference(dateFormat, timeFormat, timestampFormat));
+            return new InferSchemaAccessStrategy<>(sourceFactory, inference, getLogger());
+        }
+
+        return super.getSchemaAccessStrategy(allowableValue, schemaRegistry, context);
+    }
+
+    @Override
+    protected List<AllowableValue> getSchemaAccessStrategyValues() {
+        final List<AllowableValue> allowableValues = new ArrayList<>(super.getSchemaAccessStrategyValues());
+        allowableValues.add(HEADER_DERIVED);
+        allowableValues.add(SchemaInferenceUtil.INFER_SCHEMA);
+        return allowableValues;
+    }
+
+    @Override
+    public RecordReader createRecordReader(Map<String, String> variables, InputStream in, long inputLength, ComponentLog logger) throws MalformedRecordException, IOException, SchemaNotFoundException {
+        // Use Mark/Reset of a BufferedInputStream in case we read from the Input Stream for the header.
+        in.mark(1024 * 1024);
+        final RecordSchema schema = getSchema(variables, new NonCloseableInputStream(in), null);
+        in.reset();
+
+        ExcelRecordReaderArgs args = new ExcelRecordReaderArgs.Builder()
+                .withDateFormat(dateFormat)
+                .withDesiredSheets(desiredSheets)
+                .withFirstRow(firstRow)
+                .withInputStream(in)
+                .withLogger(logger)
+                .withSchema(schema)
+                .withTimeFormat(timeFormat)
+                .withTimestampFormat(timestampFormat)
+                .build();
+
+        return new ExcelRecordReader(args);
```

Comment:
Instead of passing the `InputStream` with the configuration, recommend keeping it as a separate argument to the constructor, the same applies to `ComponentLog`.